backend/witness/witness_test.go

```diff
@@ -114,26 +84,40 @@ func TestPublic(t *testing.T) {
 	assignment.Y = new(fr.Element).SetInt64(8000)
 	assignment.E = new(fr.Element).SetInt64(1)
 
-	w, err := New(ecc.BN254.ScalarField(), nil)
-	assert.NoError(err)
-
-	w.Schema, err = w.Vector.FromAssignment(&assignment, tVariable, false)
+	w, err := frontend.NewWitness(&assignment, ecc.BN254.ScalarField())
 	assert.NoError(err)
 
 	publicW, err := w.Public()
 	assert.NoError(err)
 
-	assert.Equal(3, w.Vector.Len())
-	assert.Equal(2, publicW.Vector.Len())
+	wt := publicW.Vector().(fr.Vector)
 
-	wt := publicW.Vector.(*witness_bn254.Witness)
+	assert.Equal(3, len(w.Vector().(fr.Vector)))
+	assert.Equal(2, len(wt))
 
-	assert.Equal("42", (*wt)[0].String())
-	assert.Equal("8000", (*wt)[1].String())
+	assert.Equal("42", wt[0].String())
+	assert.Equal("8000", wt[1].String())
 }
 
-var tVariable reflect.Type
+func roundTripMarshal(assert *require.Assertions, assignment circuit, publicOnly bool) {
+	// build the vector
+	var opts []frontend.WitnessOption
+	if publicOnly {
+		opts = append(opts, frontend.PublicOnly())
+	}
+	w, err := frontend.NewWitness(&assignment, ecc.BN254.ScalarField(), opts...)
+	assert.NoError(err)
+
+	// serialize the vector to binary
+	data, err := w.MarshalBinary()
+	assert.NoError(err)
+
+	// re-read
+	rw, err := witness.New(ecc.BN254.ScalarField())
+	assert.NoError(err)
+	err = rw.UnmarshalBinary(data)
+	assert.NoError(err)
+
+	assert.True(reflect.DeepEqual(rw, w), "witness binary round trip serialization")
 
-func init() {
-	tVariable = reflect.TypeOf(circuit{}.E)
 }
```

Comment:
I would add test case for `FromJSON`

Reply:
round trip JSON is tested in intergration_test (assert.go). Though it doesn't cost much to add something similar here too.